File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        /** 
             [1,2,4,6]
             [1,1,2,8] 
             [48,24,6,1] 
             [48,24,12,8]
        **/

        int[] preFex = new int[nums.length];
        preFex[0] = 1;

        for(int i = 1; i < nums.length; i++){
            preFex[i] = preFex[i - 1] * nums[i - 1];
        }

        int temp = 1;
        for(int i = nums.length - 1; i >= 0; i--){
            preFex[i] = temp * preFex[i];
            temp *= nums[i];
        }

        return preFex;
        
    }
}  

File: Data Structures & Algorithms/jump-game-ii/submission-2.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        """
        def dfs(jump, curr):
            if curr >= len(nums) - 1:
                return jump

            res = len(nums)
            for i in range(1, nums[curr]+1):
                res = min(res, dfs(jump+1, i + curr))
            return res
        return dfs(0, 0)
        
         0,1,2,3,4,5
        [2,4,1,1,1,1]
        [0,1,0,0,0,0]
        """

        dp = [len(nums)  for i in range(len(nums))]
        dp[0] = 0

        for i in range(1, len(dp)):
            leastJump = len(nums)
            for j in range(i):
                if j + nums[j] >= i:
                    dp[i] = min(dp[j] + 1, dp[i])

        print(dp)
        return dp[-1]


            
